Avoid loading stuff into memory to generate a flare

albertvaka opened the pull request:
### What does this PR do?

I identified a couple places where we were loading potentially big files into memory and rewrote them so this is not the case anymore.

### Motivation

Creating a flare shouldn't fail if the system is low on memory.

albertvaka committed:
Use Copy instead of ReadAll + Write

So we don't load the stacktrace into memory

albertvaka committed:
Use a pipe instead of a bytes buffer

So we don't load the entire flare zip file into memory before sending it

albertvaka committed:
The writter iface should return the bytes it processed, not written

This made a io.Copy into a RedactingWriter fail with ErrShortWrite ("write
accepted fewer bytes than requested").

Eg: io.Discard.Write returns len(buf) even though it really writes 0 bytes.

olivielpeau commented:
Nice! Haven't reviewed the change in details, but a couple of notes:

* this change deserves a release note since it's a user-facing enhancement :) (something along the lines of "reduce memory footprint of the `flare` command" would be enough)
* have you been able to profile the mem usage?

albertvaka commented:
I'm profiling right now :)

I will add a changelog 👍

truthbk commented on pkg/flare/writer.go:
```diff
@@ -95,7 +96,11 @@ func (f *RedactingWriter) Write(p []byte) (int, error) {
+		err = io.ErrShortWrite
+	}
+
+	return len(p), err
```
This is tricky from an interface implementation standpoint.... we're passing `p []byte`, but we're really writing `cleaned`, and I'm not sure all the regex cleanup will guarantee `len(p)==len(cleaned)` (probably not). Is it therefore accurate to return `len(p)` when `n != len(cleaned)` (and we have a `ShortWrite`)? I don't think that's strictly true, but not easy to get around.

albertvaka replied:
`len(p) != len(cleaned)` quite often. That's why returning `len(cleaned)` is wrong. The caller of `Write` expects to know how many bytes were _processed_ by this Writter interface, even if what was actually written is a single byte (or none, like in io.Discard). 

This is in a similar way to the `write` system call in C: as the caller, you expect it to return the length of the buffer you passed in, regardless of the actual side effects the write function has.

It's not obvious, but I asked a Go language developer about this and this is the right way.

About the extra if I added: It is just to detect the same error that Copy was detecting on us, on the underlying Writter.

truthbk replied:
I definitely prefer this new behavior, it's better and less misleading to what we previously had. 
Write processes the `p` buffer, it writes what it can, and returns the number of bytes redacted/processed (not written). And we can maybe use the error returned somehow in the called to make a decision. 

That said, this isn't strictly respecting the `Writer` interface (it was previously broken too!). Let me clarify what I mean. As you say `len(p) != len(cleaned)` is probably the most common case, as redacting the input will result in a different length byte slice. What I'm saying is that from an interface perspective that makes this `Write` function hard to use. Didn't mean to imply this was your fault, it's a design problem stemming from adding the redacting call inside the `Write()` function. Let me underline that this was broken from day 1, nothing to do with your PR.

The C `write()` syscall returns the number of bytes written:
```
RETURN VALUES
     Upon successful completion the number of bytes which were written is returned.  Otherwise, a -1 is returned and the global variable errno is set to indicate the error.
```

Similarly, the go `Write` interface is described as:
```
Write writes len(p) bytes from p to the underlying data stream. It returns the number of bytes written from p (0 <= n <= len(p)) and any error encountered that caused the write to stop early. Write must return a non-nil error if it returns n < len(p). Write must not modify the slice data, even temporarily.

type Writer interface {
        Write(p []byte) (n int, err error)
}
```

Since what we actually write is different to what the user passes, and the user has no idea what `len(redacted)` is, the value we return for bytes written would him nothing about how much of the `p` buffer passed in was written, which I believe is the goal of the return value. So I understand your change is maybe less prone to confusion, because it's an all or nothing approach - whereas the previous implementation was downright returning a meaningless number of bytes as far as the input buffer went. That's the part I meant was _"not easy to get around"_: implementing a return value that would accurately reflect the number of bytes from buffer `p` that were actually written (when writing the `redacted` buffer). Currently we say the entire `p` slice was written in the event of a short write, even if that wasn't true. And this might be the most reasonable thing we can do.

The only way I feel we could do better is if `cleaned, err := log.CredentialsCleanerBytes(p)` guaranteed `len(p) == len(cleaned)`, then we could return n.

albertvaka replied:
Just a note about C's `write()` and Go's `Write` interface: when they talk about "number of bytes which were written", it is from the caller perspective.

albertvaka committed:
Increase time we keep the connection open

Since we don't ReadAll now, but we read as we go, plus the RedactingWriter
needs to process the file as it's read, 500ms was barely enough to read
2MB on my system (with heap profiling enabled).

albertvaka committed:
Added changelong

albertvaka commented:
The new code seems fine, but with the old one I'm seeing some weird spikes of memory usage.

New code (20Mb stacktrace file like a boss):
<img width="1237" alt="Screen Shot 2019-04-03 at 17 29 20" src="https://user-images.githubusercontent.com/980842/55492491-95ec6b80-5637-11e9-8860-29b70690682a.png">

Old code (3Mb stacktrace file, see scale!):
<img width="1239" alt="Screen Shot 2019-04-03 at 17 29 29" src="https://user-images.githubusercontent.com/980842/55492502-9b49b600-5637-11e9-85d7-34caed29c005.png">

It would be nice if this can be reproduced by someone else (during QA?), because I'm quite surprised. I've made these graphs using the gcvis command. The flare file seems to come out just fine.

albertvaka committed:
Updated tests

truthbk commented on pkg/flare/flare.go:
```diff
@@ -32,41 +31,48 @@ type flareResponse struct {
+	go func() {
+
+		defer bodyWriter.Close()
+		defer writer.Close()
```
tiny nite: ideally we'd want to close `writer` before we close `bodyWriter`

albertvaka replied:
That's what we do: `defer` calls get stacked, so the last `defer` you call is the first one that gets executed.

truthbk replied:
True!

truthbk commented on pkg/flare/flare.go:
```diff
@@ -32,41 +31,48 @@ type flareResponse struct {
-	if err != nil {
-		return "", err
-	}
+	bodyReader, bodyWriter := io.Pipe()
```
We should close the `bodyReader` too.

albertvaka replied:
Good catch, thanks!

truthbk reviewed (commented):
Nice stuff with the piped writer! 👌

albertvaka committed:
Added missing Close call as per code review
